src/hooks/use-load-data-pages.ts

```diff
@@ -36,6 +37,46 @@ export const isMissingLayoutData = (layout: PivotLayout, pageInfo: PageInfo): bo
   return qWidth < Math.min(DEFAULT_PAGE_SIZE, qSize.qcx) || qHeight < Math.min(DEFAULT_PAGE_SIZE, qSize.qcy);
 };
 
+/**
+ * Note: the state of the viewService is always from the last time the chart was rendered. So it is not
+ * guaranteed to work with a new layout, as its properties may no longer be valid.
+ */
+export const getFetchArea = (
+  qLastExpandedPos: EngineAPI.INxCellPosition | undefined,
+  viewService: ViewService,
+  qSize: EngineAPI.ISize,
+  pageInfo: PageInfo,
+) => {
+  const pageStartTop = pageInfo.page * pageInfo.rowsPerPage;
+  const pageEndTop = pageStartTop + pageInfo.rowsPerPage;
+
+  if (qSize.qcy < pageStartTop) {
+    /**
+     * Do not fetch data that does not exist. This can happen because "PageInfo" is resolved from the layout.
+     * Which means it will not be updated until the next render cycle.
+     */
+    return null;
+  }
+
+  let qLeft = 0;
+  let qTop = pageStartTop;
+
+  // qLastExpandedPos only exist in the layout if a new layout was received because a node was expanded or collapsed
+  if (qLastExpandedPos) {
+    // gridColumnStartIndex might not exist anymore in the new expanded/collapsed layout
+    qLeft = Math.max(0, Math.min(qSize.qcx - DEFAULT_PAGE_SIZE, viewService.gridColumnStartIndex));
```

Comment:
why we are subtracting `DEFAULT_PAGE_SIZE` from `qSize.qcx`? 
let's look at the data from end of it in x axis:
let's say we have 1000 cols
and now we are at the last 50 col range (something like 980 as `gridColumnStartIndex`)

the `qLeft` will always endup being `qSize.qcx - 50` and always will try to fetch from index 950 when user scrolls within 950 to 1000 col. 

now if user once triggered fetched from (let's say) index 930 -> and now scrolls to the end of it's avaliable dataset (980) -> it would definitly require one more fetch to get the last 20 cols -> but this time -> it will try to fetch last 50 because of `Math.min(qSize.qcx - DEFAULT_PAGE_SIZE, viewService.gridColumnStartIndex)` 

Maybe i'm overthinking it but do we need to care about that kind of scenario? 
(Probably we have it already when data overlaps in between scrolls -> at least in y axis)

Reply:
This hook does not get called when the user scrolls. It will fetch data if the layout changes, either because of some prop change, or as in the code line, the user have expanded/collapsed a node.

Loading data on scroll is handled else where and may or may not fetch overlapping data. But that logic has not changed.